docs/cookbook/creating_custom_rest_api_response.md

```diff
@@ -0,0 +1,268 @@
+use eZ\Publish\Core\REST\Common\Output\Generator;
+use eZ\Publish\Core\REST\Common\Output\Visitor;
+use eZ\Publish\Core\REST\Server\Output\ValueObjectVisitor\VersionInfo as BaseVersionInfo;
+use eZ\Publish\API\Repository\Values\Content\VersionInfo as APIVersionInfo;
+
+class VersionInfo extends BaseVersionInfo
+{
+    private $repository;
+
+    public function __construct(Repository $repository)
+    {
+        $this->repository = $repository;
+    }
+
+    protected function visitVersionInfoAttributes(Visitor $visitor, Generator $generator, APIVersionInfo $versionInfo)
+    {
+        parent::visitVersionInfoAttributes($visitor, $generator, $versionInfo);
+
+        $this->visitTestValue($generator);
+    }
+
+    protected function visitTestValue(Generator $generator)
+    {
+        $commentsCount = 10; //for this example value is mocked, but you can use any custom logic to fetch it
+        
+        $generator->startValueElement('commentsCount', $commentsCount);
+        $generator->endValueElement('commentsCount');
+    }
+}
+```
+
+##Overriding response type
+Next, you'll need to make sure that your new implementation of serialization applies only to selected objects. In order to do that you need to 
+decorate `eZ\Publish\Core\REST\Common\Output\ValueObjectVisitorDispatcher` from `ezpublish-kernel`.
+
+```php
+<?php
+
+namespace Acme\ExampleBundle\Rest;
+
+use eZ\Publish\Core\REST\Common\Output\Generator;
+use eZ\Publish\Core\REST\Common\Output\ValueObjectVisitorDispatcher as BaseValueObjectVisitorDispatcher;
+use eZ\Publish\Core\REST\Common\Output\Exceptions\NoVisitorFoundException;
+use eZ\Publish\Core\REST\Common\Output\Visitor;
+
+class ValueObjectVisitorDispatcher extends BaseValueObjectVisitorDispatcher
+{
+    private $parentDispatcher;
+
+    public function __construct(BaseValueObjectVisitorDispatcher $parentDispatcher)
+    {
+        $this->parentDispatcher = $parentDispatcher;
+    }
+
+    public function setOutputVisitor(Visitor $outputVisitor)
+    {
+        parent::setOutputVisitor($outputVisitor);
+        $this->parentDispatcher->setOutputVisitor($outputVisitor);
+    }
+
+    public function setOutputGenerator(Generator $outputGenerator)
+    { 
+        parent::setOutputGenerator($outputGenerator);
+        $this->parentDispatcher->setOutputGenerator($outputGenerator);
+    }
+
+    public function visit($data)
+    {
+        try {
+            return parent::visit($data);
+        } catch (NoVisitorFoundException $e) {
+            return $this->parentDispatcher->visit($data);
+        }
+    }
+}
+```
+
+The response needs to have a proper type. The way to assure it is to override the format generator (e.g. xml/json).
+
+```php
+<?php
+
+namespace Acme\ExampleBundle\Rest\Generator;
+
+use eZ\Publish\Core\REST\Common\Output\Generator\Json as BaseJson;
+
+class Json extends BaseJson
+{
+    protected function generateMediaType($name, $type)
+    {
+        return "application/vnd.ez.example.{$name}+{$type}";
```

Comment:
I don't understand that part. First, it feels odd that such an override is required. Then how does it work, as it uses the same code than the extended class: 

```
    public function getMediaType($name)
    {
        return $this->generateMediaType($name, 'json');
    }

    protected function generateMediaType($name, $type)
    {
        return "application/vnd.ez.api.{$name}+{$type}";
    }
```

Please help me :)

Reply:
It allows us to identify request by our custom `Accept` header, e.g. 'application/vnd.ez.example.VersionList+json'. 

The default one doesn't allow it by using *api* part:

```
protected function generateMediaType($name, $type)
{
    return "application/vnd.ez.api.{$name}+{$type}";
}
```

In this example I was able to replace it with *example*.

Reply:
Maybe improve the naming a bit more, besides what Bertrand says maybe “my.api” as others should not use eZ as vendor name